### PR TITLE
Alg 23/logging improvements

### PR DESCRIPTION
**Pull request by matikosowy**
- Hydra experiment script
- Weight&Biases metrics logging added
- Embeddings visualization evolution animation
- Evaluation mode (pass model_path and get embeddings)
- Project paper updated (training metrics history + visualization steps comparison)
- README update
- minor fixes

### CODE REVIEW

File: midi_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        max_seq_len=cfg.train.max_seq_len,
+        tokenizer_path=tokenizer_path,
+        return_info=True,
+        limit=cfg.dataset.limit,
+    )
+    
+    # If tokenizer was null, viz_dataset trained a new one
+    tokenizer_path = viz_dataset.tokenizer_path 
+    
+    if hasattr(cfg, "eval") and cfg.eval.enabled:
```
Comment (WojciechMat) [inline midi_embeddings.py:39]: There is a return in the if block, so some ~50 lines of code in this script are always ommitted (the if block itself or everything after it). Would it make sense to split this into separate scripts?

train.py - Just for training
evaluate.py - Just for evaluation/visualization
main.py - A thin wrapper that calls the correct script if you still want a single entry point.

This would make each file have one clear job instead of a too big conditional block handling different pipelines.

File: midi_embeddings/visualize.py [hunk 2]
```diff
@@ -179,13 +238,124 @@ def visualize_embeddings(
...
+        for i in range(len(self.dataset)):
+            info = self.dataset[i]["info"]
+            titles.append(info["title"])
+            composers.append(info.get("composer", "Unknown"))
+
+        raw_embeddings = np.vstack(all_embeddings)
+
+        tsne = TSNE(random_state=42)
+        embeddings_2d = tsne.fit_transform(raw_embeddings)
+
```
Comment (WojciechMat) [inline midi_embeddings/visualize.py:282]: That's cool! but very unstable on visualizations, because you are re-fitting the estimator each time.

I suggest using an estimator that you can fit to the embedding once and use at all times later. For example, if you use PCA, you can do `pca.fit(...)` and then, for each embedding, do' pca.transform(...)` to get the 2D embedding. 

It would make sense to fit PCA on embeddings that you get after the training and then use the same estimator for embeddings saved during training.

**Review by WojciechMat — APPROVED** [review]
Great work! Beautifully integrated with hydra, bonus points for plots on wandb 🚀 🚀 🚀 
Left some comments too 😸
